File: lib-commons/src/main/java/com/epam/learn/microservices/fundamentals/logging/LoggingAspect.java
package com.epam.learn.microservices.fundamentals.logging;

import com.epam.learn.microservices.fundamentals.logging.LogExecution.Level;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.AfterThrowing;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.MethodSignature;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Aspect
public class LoggingAspect {

    @Around(value = "executionLogging()")
    public Object logMethodExecution(final ProceedingJoinPoint joinPoint) throws Throwable {
        var annotation = findAnnotation(joinPoint);

        if (annotation != null) {
            var logger = componentLogger(joinPoint);
            var level = annotation.level();

            if (isLevelEnabled(logger, level)) {
                var signature = joinPoint.getSignature();
                var signatureName = "%s.%s()".formatted(
                    signature.getDeclaringType().getSimpleName(),
                    signature.getName()
                );

                log(logger, level, "Method: {}, Args: {}", signatureName, joinPoint.getArgs());
                var result = joinPoint.proceed();
                log(logger, level, "Method: {}, Result: {}", signatureName, result);

                return result;
            }
        }
        return joinPoint.proceed();
    }

    @AfterThrowing(pointcut = "executionLogging()", throwing = "exception")
    public void logException(final JoinPoint joinPoint, final Throwable exception) {
        componentLogger(joinPoint).error("An error occurred", exception);
    }

    @Pointcut("within(@com.epam.learn.microservices.fundamentals.logging.LogExecution *)")
    private void annotatedClass() { /* no-op */ }

    @Pointcut("@annotation(com.epam.learn.microservices.fundamentals.logging.LogExecution)")
    private void annotatedMethod() { /* no-op */ }

    @Pointcut("execution(public * *(..))")
    private void publicMethod() { /* no-op */ }

    @Pointcut("(annotatedClass() || annotatedMethod()) && publicMethod()")
    private void executionLogging() { /* no-op */ }

    private Logger componentLogger(final JoinPoint joinPoint) {
        return LoggerFactory.getLogger(joinPoint.getSignature().getDeclaringTypeName());
    }

    private LogExecution findAnnotation(final JoinPoint joinPoint) {
        if (joinPoint.getSignature() instanceof MethodSignature signature) {
            var annotationClass = LogExecution.class;
            var method = signature.getMethod();
            var annotation = method.getAnnotation(annotationClass);

            if (annotation == null) {
                var declaringClass = method.getDeclaringClass();

                annotation = declaringClass.getAnnotation(annotationClass);
            }
            return annotation;
        }
        return null;
    }

    private boolean isLevelEnabled(final Logger logger, final Level level) {
        return switch (level) {
            case DEBUG -> logger.isDebugEnabled();
            case INFO -> logger.isInfoEnabled();
        };
    }

    private void log(final Logger logger, final Level level, final String msg, final Object... ars) {
        switch (level) {
            case DEBUG -> logger.debug(msg, ars);
            case INFO -> logger.info(msg, ars);
        }
    }
}

File: lib-commons/src/main/kotlin/com/epam/learn/microservices/fundamentals/logging/LoggingAspect.kt
package com.epam.learn.microservices.fundamentals.logging

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.AfterThrowing
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.slf4j.Logger
import org.slf4j.LoggerFactory

@Aspect
class LoggingAspect {

    @Around("executionLogging()")
    fun logMethodExecution(joinPoint: ProceedingJoinPoint, annotation: LogExecution): Any? {
        val logger = joinPoint.componentLogger

        return if (logger.isLevelEnabled(annotation.level)) {
            val signature = joinPoint.signature
            val signatureName = "${signature.declaringType.simpleName}.${signature.name}()"

            logger.log(annotation.level, "Method: {}, Args: {}", signatureName, joinPoint.args.joinToString())
            val result = joinPoint.proceed()
            logger.log(annotation.level, "Method: {}, Result: {}", signatureName, result)

            result
        } else {
            joinPoint.proceed()
        }
    }

    @AfterThrowing(pointcut = "executionLogging()", throwing = "exception")
    fun logException(joinPoint: JoinPoint, exception: Throwable) {
        joinPoint.componentLogger.error("An error occurred", exception)
    }

    @Pointcut("within(@com.epam.learn.microservices.fundamentals.logging.LogExecution *)")
    private fun annotatedClass() {
        /* no-op */
    }

    @Pointcut("@annotation(com.epam.learn.microservices.fundamentals.logging.LogExecution)")
    private fun annotatedMethod() {
        /* no-op */
    }

    @Pointcut("execution(public * *(..))")
    private fun publicMethodExecution() {
        /* no-op */
    }

    @Pointcut("(annotatedClass() || annotatedMethod()) && publicMethodExecution()")
    private fun executionLogging() {
        /* no-op */
    }

    private val JoinPoint.componentLogger: Logger
        get() = LoggerFactory.getLogger(signature.declaringTypeName)

    private fun Logger.isLevelEnabled(level: LogExecution.Level): Boolean {
        return when (level) {
            LogExecution.Level.DEBUG -> isDebugEnabled
            LogExecution.Level.INFO -> isInfoEnabled
        }
    }

    private fun Logger.log(level: LogExecution.Level, format: String, vararg arguments: Any) {
        when (level) {
            LogExecution.Level.DEBUG -> debug(format, *arguments)
            LogExecution.Level.INFO -> info(format, *arguments)
        }
    }
}

